src/query/functions/src/scalars/map.rs

```diff
@@ -244,4 +252,121 @@ pub fn register(registry: &mut FunctionRegistry) {
                     .any(|(k, _)| k == key)
             },
         );
+
+    registry.register_3_arg_core::<EmptyMapType, GenericType<0>, GenericType<1>, MapType<GenericType<0>, GenericType<1>>, _, _>(
+        "map_insert",
+        |_, _, insert_key_domain, insert_value_domain| {
+            FunctionDomain::Domain(Some((
+                insert_key_domain.clone(),
+                insert_value_domain.clone(),
+            )))
+        },
+        |_, key, value, ctx| {
+            let mut b = ArrayType::create_builder(1, ctx.generics);
+            b.put_item((key.into_scalar().unwrap(), value.into_scalar().unwrap()));
+            b.commit_row();
+            return Value::Scalar(MapType::build_scalar(b));
+        },
+    );
+
+    registry.register_passthrough_nullable_3_arg(
+        "map_insert",
+        |_, domain1, key_domain, value_domain| {
+            FunctionDomain::Domain(match (domain1, key_domain, value_domain) {
+                (Some((key_domain, val_domain)), insert_key_domain, insert_value_domain) => Some((
+                    key_domain.merge(insert_key_domain),
+                    val_domain.merge(insert_value_domain),
+                )),
+                (None, _, _) => None,
+            })
+        },
+        vectorize_with_builder_3_arg::<
+            MapType<GenericType<0>, GenericType<1>>,
+            GenericType<0>,
+            GenericType<1>,
+            MapType<GenericType<0>, GenericType<1>>,
+        >(|source, key, value, output, ctx| {
+            // insert operation only works on specific key type: boolean, string, numeric, decimal, date, datetime
+            let key_type = &ctx.generics[0];
+            if !key_type.is_boolean()
```

Comment:
move this check to `EmptyMap` function, the key type of map is valid, don't need to check.